#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
df = pd.read_csv('drive/MyDrive/Date_Analysis/data/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [4]:
df.groupby('Branch')['Total'].sum()

,Total
Branch,
A,106200.3705
B,106197.6720
C,110568.7065


Філіал С має найвищий прибуток.

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [5]:
df[df.Branch == 'C'].City.iloc[0]

'Naypyitaw'

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [6]:
invoices_by_category = pd.pivot_table(df, index='Branch', columns='Product line', values= 'Invoice ID', aggfunc='count')

In [7]:
invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [8]:
invoices_by_category['Electronic accessories'].nlargest(1)

,Electronic accessories
Branch,
A,60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**Питання № 1** - Товари яких категорій найбільше / найменше купуть жінки, а яких чоловіки?

In [26]:

pd.pivot_table(df, index='Gender', columns='Product line', values='Invoice ID', aggfunc='count')

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Gender,,,,,,
Female,84,96,90,64,79,88
Male,86,82,84,88,81,78


Бачимо, що жінки найбільше купують товари категорії 'Модні аксесуари', а найменше - 'Здоров'я та краса'. Чоловіки ж, навпаки, надають перевагу категорії 'Здоров'я та краса', а найменше - 'Спорт та подорожі' (що є доволі неочікуваним).

То ж можна, наприклад, закупати більше категорій товарів, які продаються найкраще. Також, можна розмістити рекламу з товарами, які продаються гірше.


**Питання № 2** -  Який тип клієнтів дає нам більший валовий дохід?

In [40]:
df.groupby(by='Customer type')['gross income'].agg('sum')

,gross income
Customer type,
Member,7820.164
Normal,7559.205


Більший валовий дохід магазини отримують від клієнтів, які мають картку лояльності клієнта. Тобто сенс в програмі лояльності є, але можливо ми можемо ще більше покращити її? До прикладу, ввести систему накопичення бонусів, або знижки на окремі види товарів.


**Питання № 3** - Який середній рейтинг магазинів по містам?

In [36]:
df.groupby(by='City')['Rating'].mean()

,Rating
City,
Mandalay,6.818072
Naypyitaw,7.072866
Yangon,7.027059


В місті Мандалай магазин з найнижчим рейтингом - 6.8. Ц змушує замислитись, чи впливає рівень рейтингу на прибуток в філії цього міста. Можливо, варто провести перевірку і розібратись в причинах такого невисокого рейтингу.